src/site/stages/build/plugins/check-broken-links/index.js

```diff
@@ -0,0 +1,54 @@
+/* eslint-disable no-param-reassign, no-console, no-continue */
+
+const path = require('path');
+
+const ENVIRONMENTS = require('../../../../constants/environments');
+
+const _getBrokenLinks = require('./helpers/getBrokenLinks');
+const _applyIgnoredRoutes = require('./helpers/applyIgnoredRoutes');
+const _getErrorOutput = require('./helpers/getErrorOutput');
+
+function getMiddleware(
+  buildOptions,
+  getBrokenLinks = _getBrokenLinks,
+  applyIgnoredRoutes = _applyIgnoredRoutes,
+  getErrorOutput = _getErrorOutput,
+) {
+  return (files, metalsmith, done) => {
+    const fileNames = Object.keys(files);
+    const allPaths = new Set(fileNames);
+
+    let brokenPages = [];
+
+    for (const fileName of fileNames) {
+      const isHtml = path.extname(fileName) === '.html';
+      if (!isHtml) continue;
+
+      const file = files[fileName];
+      const linkErrors = getBrokenLinks(file, allPaths);
+
+      if (linkErrors.length > 0) {
+        brokenPages.push({
+          path: file.path,
+          linkErrors,
+        });
+      }
+    }
+
+    brokenPages = applyIgnoredRoutes(brokenPages, files);
+
+    if (brokenPages.length > 0) {
+      const errorOutput = getErrorOutput(brokenPages);
+
+      if (buildOptions.buildtype === ENVIRONMENTS.VAGOVPROD) {
+        done(errorOutput);
+        return;
+      }
+      console.log(errorOutput);
+    }
+
+    done();
+  };
+}
+
+module.exports = getMiddleware;
```

Comment:
This is the middleware entry point. I restructured the single `check-broken-links.js` file into a directory because the individual functions were complex so worth adding unit tests for, and it was cleaner to test when broken out into separate files.

This file ultimately -

1. Loops through all of the Metalsmith files
2. Executes `getBrokenLinks` on each HTML page, forming a data structure out of the pages that returned arrays link errors
3. If there are link-errors, it formats the link-errors into useful output. It will break the build on production, but just log the errors on lower environments.
4. If no link-errors, then the build carries on without anything being logged.